Sources/SwiftBasicFormat/BasicFormat.swift

```diff
@@ -0,0 +1,278 @@
+    }()
+
+    lazy var nextTokenWillStartWithNewline: Bool = {
+      guard let nextToken = token.nextToken(viewMode: .sourceAccurate) else {
+        return false
+      }
+      return nextToken.leadingTrivia.first?.isNewline ?? false
+        || requiresLeadingNewline(nextToken)
+    }()
+
+    lazy var trailingTriviaAndNextTokensLeadingTriviaWhitespace: Trivia = {
+      let nextToken = token.nextToken(viewMode: .sourceAccurate)
+      let nextTokenLeadingWhitespace = nextToken?.leadingTrivia.prefix(while: { $0.isIndentationWhitespace }) ?? []
+      return trailingTrivia + Trivia(pieces: nextTokenLeadingWhitespace)
+    }()
+
+    var leadingTrivia = token.leadingTrivia
+    var trailingTrivia = token.trailingTrivia
+
+    if requiresLeadingNewline(token) {
+      // Add a leading newline if the token requires it unless
+      //  - it already starts with a newline or
+      //  - the previous token ends with a newline
+      if !(leadingTrivia.first?.isNewline ?? false) && !previousTokenWillEndInNewline {
+        // Add a leading newline if the token requires it and
+        //  - it doesn't already start with a newline and
+        //  - the previous token didn't end with a newline
+        leadingTrivia = .newline + leadingTrivia
+      }
+    } else if requiresLeadingBlank(token) {
+      // Add a leading space if the token requires it unless
+      //  - it already starts with a blank or
+      //  - the previous token ends with a blank after the rewrite
+      if !(leadingTrivia.first?.isBlank ?? false) && !previousTokenWillEndWithBlank {
+        leadingTrivia += .space
+      }
+    }
+
+    if leadingTrivia.indentation(isOnNewline: previousTokenWillEndInNewline) == [] {
+      // If the token starts on a new line and does not have indentation, this
+      // is the last non-indented token. Store its indentation level
+      anchorPoints[token] = currentIndentationLevel
+    }
+
+    // Add a trailing space to the token unless
+    //  - it already ends with a blank or
+    //  - the next token will start starts with a newline after the rewrite
+    //    because newlines should be preferred to spaces as a blank
+    if requiresTrailingBlank(token)
+      && !(trailingTrivia.pieces.last?.isBlank ?? false)
+      && !nextTokenWillStartWithNewline
+    {
+      trailingTrivia += .space
+    }
+
+    var leadingTriviaIndentation = self.currentIndentationLevel
+    var trailingTriviaIndentation = self.currentIndentationLevel
+
+    // If the trivia contain user-defined indentation, find their anchor point
+    // and indent the token relative to that anchor point.
+    if leadingTrivia.containsIndentation(isOnNewline: previousTokenWillEndInNewline),
+      let anchorPointIndentation = self.anchorPointIndentation(for: token)
+    {
+      leadingTriviaIndentation = anchorPointIndentation
+    }
+    if trailingTriviaAndNextTokensLeadingTriviaWhitespace.containsIndentation(isOnNewline: previousTokenWillEndInNewline),
```

Comment:
Do we really care about the trailing? There shouldn't be a newline there anyway. That would also let us get rid of all the `isOnNewline`.

Reply:
We have newlines as trailing trivia for multi-line string literals. In that case the newline after the `"""` is trailing trivia of the quotes.